feat: add AS5600 component/sensor

ammmze opened the pull request:
The AS5600 is a magnetic rotary position encoder with a 12-bit resolution, allowing for 4096 positions/steps. There's been an open feature request for this for a couple years and a custom component implementation using a 3rd party library. This component uses the esphome native i2c implementation, not the 3rd party library.

resolves: esphome/feature-requests#1459

# What does this implement/fix?

The AS5600 is a magnetic rotary position encoder with a 12-bit resolution, allowing for 4096 positions/steps. There's been an open feature request for this for a couple years and a custom component implementation using a 3rd party library. This component uses the esphome native i2c implementation, not the 3rd party library.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

**Related issue or feature (if applicable):** resolves esphome/feature-requests#1459

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3304

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
i2c:
  sda: 25
  scl: 21
  scan: true
  id: bus_a

as5600:
    id: my_as5600
    dir_pin: 22
    start_position: -90deg
    range: 180deg

sensor:
  - platform: as5600
    update_interval: 1s
    name: Percent
    unit_of_measurement: '%'
    accuracy_decimals: 2
    icon: mdi:percent
    filters:
      - delta: 1
      - lambda: 'return (x / as5600::POSITION_COUNT) * 100;'
  - platform: as5600
    update_interval: 1s
    name: Angle
    unit_of_measurement: °
    accuracy_decimals: 2
    icon: mdi:rotate-right
    out_of_range_mode: nan
    filters:
      - delta: 1
      - lambda: 'return x * as5600::RAW_TO_DEGREES * id(my_as5600).get_range_scale();'
  - platform: as5600
    update_interval: 1s
    name: Position
    filters:
      - delta: 1
    raw_position:
      name: Raw Position
      filters:
        - delta: 1
    gain:
      name: Gain
      filters:
        - delta: 1
    magnitude:
      name: Magnitude
      filters:
        - delta: 1
    status:
      name: Status
      filters:
        - delta: 1

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

ammmze committed:
feat: add AS5600 component/sensor

The AS5600 is a magnetic rotary position encoder with a 12-bit resolution, allowing for 4096 positions/steps. There's been an open feature request for this for a couple years and a custom component implementation using a 3rd party library. This component uses the esphome native i2c implementation, not the 3rd party library.

resolves: esphome/feature-requests#1459
Signed-off-by: Branden Cash <203336+ammmze@users.noreply.github.com>

fran6120 commented:
Hi!

So far I was using this sensor in esphome by integrating as an external component from here: https://github.com/esphome/feature-requests/issues/1459

Now, seeing that you have written a custom library for esphome, I have been encouraged to try it.

The utility that I give is to obtain the direction in degrees of a 3D printed weather vane and for the moment it works great.

The only problem that I have found for my use, is that the lambda function that returns the value of the position in degrees, sends both the value "0 deg" and "360 deg". This problem already happened with the previous external component.

I don't think it's the most elegant way to do it, but it can be solved like this:

```
 - lambda: |-
       float deg = floor (x * as5600::RAW_TO_DEGREES * id(my_as5600).get_range_scale());
       if (deg > 359) {return deg - 360;}
       else {return deg;}
```

Perhaps it would be more convenient to have the option of obtaining the sensor value in degrees or crude, in the sensor configuration itself without having to use a lambda, but only as a suggestion, since it works great as it is.

Thanks you!

ammmze commented:
Thank you for your feedback! It is greatly appreciated!

> The only problem that I have found for my use, is that the lambda function that returns the value of the position in degrees, sends both the value "0 deg" and "360 deg".

That is interesting that you would get 360 deg because `x` should be a value equal to or between `0` and `4095`. So a max degree _should_ come through as `359.912109375` (eq `4095 * (360 / 4096)`). And then using `floor`, in theory that should come out to 359. I will look more into this.

Also for a slightly more elegant solution you could use the modulo operator. Replace the last 2 lines of your lambda with `return deg % 360;`.

Some examples of this:
```
  0 % 360 = 0
  1 % 360 = 1
180 % 360 = 180
359 % 360 = 359
360 % 360 = 0
361 % 360 = 1
```

> Perhaps it would be more convenient to have the option of obtaining the sensor value in degrees or raw, in the sensor configuration itself without having to use a lambda, but only as a suggestion, since it works great as it is.

That is something I am open to doing. I had thought about doing that but I wasn't sure if that is common in these esphome components. I haven't really come across other components that give an option to change the reporting format, that was really the primary reason why I didn't implement it. Although then if we do make an option to have it convert to degrees within the component, we'd then need to decide does it floor the degrees to give a whole degree, does it round to a certain decimal place, etc. The use case for any of them could be valid...this is probably a contributing reason for why we don't really see other components doing options for conversion built-in to the component.

fran6120 commented:
You are right! only using `floor` solve the 0 and 360 issue.

The other part of the code that no longer works, was used with the previous library, when adding an angle start offset.  Function that you have implemented with `start_position`

Thanks you!

jesserockz commented on esphome/components/as5600/as5600.cpp:
```diff
@@ -0,0 +1,186 @@
+                                                          : (4095 - this->start_position_) + this->raw_max_;
+
+  // range scale is ratio of actual allowed range to the full range
+  this->range_scale_ = range / 4095.0;
```
```suggestion
  this->range_scale_ = range / 4095.0f;
```

ammmze replied:
👍🏻  Done!

jesserockz commented on esphome/components/as5600/__init__.py:
```diff
@@ -0,0 +1,229 @@
+
+CODEOWNERS = ["@ammmze"]
+DEPENDENCIES = ["i2c"]
+AUTO_LOAD = ["sensor"]
```
Please alter the PR code to work without autoloading `sensor`
```suggestion
```

ammmze replied:
👍🏻 Done! Good catch...guess I just forgot to remove that after I moved the sensors to `sensor.py`.

ammmze commented:
Looks like my [esphome-docs PR](https://github.com/esphome/esphome-docs/pull/3098) got closed and locked since the PRs have just been sitting idle for so long waiting for reviews. I've just rebased it and pushed new changes. Are you, @jesserockz, able to unlock it and re-open it? Or do I need to open a new PR in esphome-docs?

jesserockz commented:
@ammmze GitHub wont let me reopen it because you force pushed your branch afterwards. Please make a new one.

jesserockz commented on esphome/components/as5600/sensor.py:
```diff
@@ -0,0 +1,120 @@
+from . import as5600_ns, AS5600Component
+
+CODEOWNERS = ["@ammmze"]
+DEPENDENCIES = ["as5600"]
+AUTO_LOAD = ["as5600"]
```
This should be one or the other, not both.

jesserockz reviewed (changes requested):
Please move the sensor class to its own file `as5600_sensor.h/cpp` and place it in a `sensor` subdirectory. Also rename `sensor.py` to be `sensor/__init__.py` in that new subdirectory. 
This will also require altering some of the main component code as it should not know about AS5600Sensor. This is something that many ESPHome components do already if you need examples.

ammmze committed:
refactor: move sensor to sub-package

ammmze commented:
> Please move the sensor class to its own file `as5600_sensor.h/cpp` and place it in a `sensor` subdirectory. Also rename `sensor.py` to be `sensor/__init__.py` in that new subdirectory. This will also require altering some of the main component code as it should not know about AS5600Sensor. This is something that many ESPHome components do already if you need examples.

These changes have been completed and I've opened a [new PR in esphome-docs](https://github.com/esphome/esphome-docs/pull/3304) to replace the one that got closed.

jesserockz commented on esphome/components/as5600/sensor/__init__.py:
```diff
@@ -0,0 +1,119 @@
+    await cg.register_component(var, config)
+    await sensor.register_sensor(var, config)
+
+    if CONF_OUT_OF_RANGE_MODE in config:
+        cg.add(var.set_out_of_range_mode(config[CONF_OUT_OF_RANGE_MODE]))
```
We have been slowly converting ESPHome to_code functions to use this format for optional config entries. Please alter them to match =)
```suggestion
    if out_of_range_mode_config := config.get(CONF_OUT_OF_RANGE_MODE):
        cg.add(var.set_out_of_range_mode(out_of_range_config_mode))
```

jesserockz commented on esphome/components/as5600/sensor/as5600_sensor.cpp:
```diff
@@ -0,0 +1,79 @@
+
+float AS5600Sensor::get_setup_priority() const { return setup_priority::DATA; }
+
+void AS5600Sensor::dump_config() {
+  LOG_SENSOR("", "AS5600 Sensor", this);
+  ESP_LOGCONFIG(TAG, "  Out of Range Mode: %u", this->out_of_range_mode_);
+  LOG_UPDATE_INTERVAL(this);
+}
```
This should also run LOG_SENSOR for all of the sub-sensors

jesserockz commented on esphome/components/as5600/sensor/as5600_sensor.h:
```diff
@@ -0,0 +1,43 @@
+
+class AS5600Sensor : public PollingComponent, public Parented<AS5600Component>, public sensor::Sensor {
+ public:
+  AS5600Sensor();
+
```
Not doing anything useful
```suggestion
```

jesserockz commented on esphome/components/as5600/sensor/as5600_sensor.cpp:
```diff
@@ -0,0 +1,79 @@
+static const uint8_t REGISTER_AGC = 0x1A;        // 8 bytes  / R
+static const uint8_t REGISTER_MAGNITUDE = 0x1B;  // 16 bytes / R
+
+AS5600Sensor::AS5600Sensor(){};
+
```
```suggestion
```

jesserockz commented on esphome/components/as5600/sensor/as5600_sensor.h:
```diff
@@ -0,0 +1,43 @@
+  void dump_config() override;
+  float get_setup_priority() const override;
+
+  void set_angle_sensor(sensor::Sensor *angle_sensor) { angle_sensor_ = angle_sensor; }
```
Please make sure all member variables are prefixed with `this->` as per our [codebase standards](https://esphome.io/guides/contributing#codebase-standards)

```suggestion
  void set_angle_sensor(sensor::Sensor *angle_sensor) { this->angle_sensor_ = angle_sensor; }
```

jesserockz reviewed (changes requested):
Thanks for making the requested changes and splitting the sensor code out.

A few more things below, but nothing serious =)

ammmze committed:
chore: apply requested PR changes

chore: update to_code config to match latest code style

see: https://github.com/esphome/esphome/pull/5174#discussion_r1382749817

chore: remove useless method defs

see: https://github.com/esphome/esphome/pull/5174#discussion_r1382750138
see: https://github.com/esphome/esphome/pull/5174#discussion_r1382750191

chore: ensure member vars are accessed via this->

see: https://github.com/esphome/esphome/pull/5174#discussion_r1382750916

feat: log all enabled child sensors

see: https://github.com/esphome/esphome/pull/5174#discussion_r1382749995

ammmze commented:
> Thanks for making the requested changes and splitting the sensor code out.
> 
> A few more things below, but nothing serious =)

Thank you for taking the time to look through this again! I've made the requested changes.

jesserockz commented on esphome/components/as5600/__init__.py:
```diff
@@ -0,0 +1,228 @@
+        pin = await cg.gpio_pin_expression(dir_pin_config)
+        cg.add(var.set_dir_pin(pin))
+
+    if end_position_config := config.get(CONF_END_POSITION):
+        cg.add(var.set_end_position(end_position_config))
+
+    if range_config := config.get(CONF_RANGE):
+        cg.add(var.set_range(range_config))
```
These need to have `None` checks as they could be 0 integers which would be falsy and be skipped over.
```suggestion
    if (end_position_config := config.get(CONF_END_POSITION, None)) is not None:
        cg.add(var.set_end_position(end_position_config))

    if (range_config := config.get(CONF_RANGE, None)) is not None:
        cg.add(var.set_range(range_config))
```

jesserockz committed:
Update esphome/components/as5600/__init__.py

jesserockz commented:
Dont forget to take a look at the docs PR changes. Once both PRs are approved they can be merged =)
